test/unit/utils.test.js

```diff
@@ -456,4 +458,92 @@ describe('utils', function () {
       });
     });
   });
+
+  describe('shuffle function', () => {
+    it('should support iterables', function () {
+      // Kind of an implicit test, we should not throw/crash here.
+      const input = new Set(['a', 'b', 'c']);
+      const output = shuffle(input);
+      expect(Array.isArray(output)).to.be.true;
+    });
+
+    it('should not mutate the original input', function () {
+      const input = Object.freeze(['a', 'b', 'c', 'd', 'e']);
+      const output = shuffle(input); // This will throw if shuffle tries to edit the input
+      expect(output).to.not.deep.equal(input);
+      expect(output).to.have.lengthOf(input.length);
+    });
+
+    it(`should give a random subset of length input.length - 1`, function () {
+      const input = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i'];
+      const output = shuffle(input, input.length - 1);
+      expect(output).to.not.deep.equal(input);
+      expect(output).to.have.lengthOf(input.length - 1);
+    });
+
+    it(`should give a random subset of length input.length`, function () {
+      const input = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i'];
+      const output = shuffle(input, input.length);
+      expect(output).to.not.deep.equal(input);
+      expect(output).to.have.lengthOf(input.length);
+    });
+
+    it(`should always return the same element when input is one item`, function () {
+      const input = ['a'];
+      for (let i = 0; i < 10; i++) {
+        const output = shuffle(input);
+        expect(output).to.deep.equal(input);
+      }
+      for (let i = 0; i < 10; i++) {
+        const output = shuffle(input, 1); // and with limit
+        expect(output).to.deep.equal(input);
+      }
+    });
+
+    it(`should return a different item on every call of limit 1`, function () {
```

Comment:
small technicality, since we can't guarantee it'll be "different" on any given call
```suggestion
    it(`should return a random item on every call of limit 1`, function () {
```